### PR TITLE
V3 swagger epankou

### PR DESCRIPTION
**Pull request by epankou**

### CODE REVIEW

File: specs/swagger.yml [hunk 1]
```diff
@@ -442,47 +441,77 @@ paths:
...
             X-Timestamp:
               type: number
               description: FogController server timestamp
-        401:
+        '401':
           description: Not Authorized
-        404: 
+        '404':
+          description: Invalid Node Id
+  '/iofog/{id}/agent/proxy/config':
```
Comment (baghbidi) [inline specs/swagger.yml]: GET `/iofog/{id}/agent/proxy` or `/iofog/{id}/agent/proxy-config` would be better imo.

File: specs/swagger.yml [hunk 2]
```diff
@@ -969,26 +998,147 @@ paths:
...
-        401:
+        '401':
           description: Not Authorized
-        404:
+        '404':
           description: Not Found
-      
-
-
+  '/iofog/proxy':
```
Comment (baghbidi) [inline specs/swagger.yml]: POST `iofog/{id}/proxy`. then request body is not required.

File: specs/swagger.yml [hunk 3]
```diff
@@ -969,26 +998,147 @@ paths:
...
+        - in: body
+          name: fogNodeInfo
+          required: true
+          schema:
+            $ref: '#/definitions/CreateIOFoNodeProxyRequestBody'
+      responses:
+        '201':
+          description: Created
+          schema:
+            $ref: '#/definitions/IOFogNodeProxyResponse'
```
Comment (baghbidi) [inline specs/swagger.yml]: response is not required. user already has the fogId (passed as path parameter)

File: specs/swagger.yml [hunk 4]
```diff
@@ -969,26 +998,147 @@ paths:
...
+          headers:
+            X-Timestamp:
+              type: number
+              description: FogController server timestamp
+        '400':
+          description: Bad Request
+        '401':
+          description: Not Authorized
```
Comment (baghbidi) [inline specs/swagger.yml]: if you use `iofog/{id}/proxy`, add an 404 error as well.

File: specs/swagger.yml [hunk 5]
```diff
@@ -969,26 +998,147 @@ paths:
...
+  '/iofog/proxy/{id}':
```
Comment (baghbidi) [inline specs/swagger.yml]: `iofog/{id}/proxy`.
Move `patch` and `get` under previous path 👆
Also, don't we need a `delete` for proxy?

File: specs/swagger.yml [hunk 6]
```diff
@@ -969,26 +998,147 @@ paths:
...
+            $ref: '#/definitions/IOFogNodeProxyInfoResponse'
+          headers:
+            X-Timestamp:
+              type: number
+              description: FogController server timestamp
+        '401':
+          description: Not Authorized
+        '404':
+          description: Invalid Node Id
+  '/iofog/proxy/{id}/status':
```
Comment (baghbidi) [inline specs/swagger.yml]: `iofog/{id}/proxy/status`

File: specs/swagger.yml (hunk 3 above)
Comment (epankou) [reply specs/swagger.yml]: fixed

File: specs/swagger.yml [hunk 7]
```diff
@@ -969,30 +1009,231 @@ paths:
...
+            $ref: '#/definitions/IOFogNodeTunnelStatusResponse'
+          headers:
+            X-Timestamp:
+              type: number
+              description: FogController server timestamp
+        '401':
+          description: Not Authorized
+        '404':
+          description: Invalid Node Id
+  '/registry/list':
```
Comment (baghbidi) [inline specs/swagger.yml]:
```
/registries
  get: ...
  post: ...
/registries/{id}:
  get: ...
  patch: ...
  delete: ...
```

File: specs/swagger.yml [hunk 8]
```diff
@@ -969,30 +1009,231 @@ paths:
...
         type: string
         example: ok
       timestamp:
         type: number
+  EmailActivationStatusResponse:
+    type: object
+    properties:
+      emailActivation:
+        type: string
+      timestamp:
```
Comment (baghbidi) [inline specs/swagger.yml]: `timestamp` is not necessary. All endpoints return server time as `X-Timestamp` in response header

File: specs/swagger.yml [hunk 9]
```diff
@@ -1029,9 +1272,9 @@ definitions:
       description:
         type: string
       createAt:
-        type: number
```
Comment (baghbidi) [inline specs/swagger.yml]: This should be `number` as well as `updatedAt` and `microserviceMessageCounts`. Type casting must be done in ioFog agent.

File: specs/swagger.yml [hunk 10]
```diff
@@ -1096,21 +1339,39 @@ definitions:
...
-      configFrequency:
-        type: number      
+        type: number
+      changeFrequency:
+        type: number
+      scanFrequency:
+        type: number
+      proxy:
+        type: string
       typeKey:
```
Comment (baghbidi) [inline specs/swagger.yml]: I think it's better to have `type` as a JSON property here.
```
type:
  type: object
  properties:
    key:
      type: number
    id:
      type: number
    name:
      type: string
    image:
      type: string
    description:
      type: string
```

File: specs/swagger.yml [hunk 11]
```diff
@@ -1372,9 +1650,28 @@ definitions:
         type: string
       password:
         type: string
+      useremail:
```
Comment (baghbidi) [inline specs/swagger.yml]: `userEmail`

File: specs/swagger.yml [hunk 12]
```diff
@@ -1372,9 +1650,28 @@ definitions:
...
+        type: string
+      user_id:
```
Comment (baghbidi) [inline specs/swagger.yml]: `userId`

File: specs/swagger.yml [hunk 13]
```diff
@@ -1372,9 +1650,28 @@ definitions:
...
+        type: string
+      iofog_uuid:
```
Comment (baghbidi) [inline specs/swagger.yml]: `ioFogId`

File: specs/swagger.yml (hunk 10 above)
Comment (epankou) [reply specs/swagger.yml]: do you mean fogType? 
like
fogType:
  type: number

File: specs/swagger.yml [hunk 14]
```diff
@@ -969,30 +1093,377 @@ paths:
...
+          schema:
+            type: object
+            properties:
+              id:
+                type: string
+          headers:
+            X-Timestamp:
+              type: number
+              description: FogController server timestamp
+        '400':
```
Comment (baghbidi) [inline specs/swagger.yml]: not required!

File: specs/swagger.yml [hunk 15]
```diff
@@ -969,30 +1093,377 @@ paths:
...
+          description: Success
+          headers:
+            X-Timestamp:
+              type: number
+              description: FogController server timestamp
+        '401':
+          description: Not Authorized
+        '404':
+          description: Invalid Microservice Id
+  '/iofog/microservices/{id}/strace/json':
```
Comment (baghbidi) [inline specs/swagger.yml]:
```
/iofog/microservices/{id}/strace:
  get:
    parameters:
      in: query
      name: format
      required: true
      type: string
      enum:
        - file
        - string
```

File: specs/swagger.yml [hunk 16]
```diff
@@ -969,30 +1093,377 @@ paths:
...
+                type: string
+          headers:
+            X-Timestamp:
+              type: number
+              description: FogController server timestamp
+        '401':
+          description: Not Authorized
+        '404':
+          description: Invalid Microservice Id
+  '/iofog/microservices/{id}/strace/ftp':
```
Comment (baghbidi) [inline specs/swagger.yml]: `PUT /iofog/microservices/{id}/strace`

**Pull request merged**